ledger/eval.go

```diff
@@ -697,14 +656,100 @@ func (eval *BlockEvaluator) GenerateBlock() (*ValidatedBlock, error) {
 	return &vb, nil
 }
 
+type evalTxValidator struct {
+	txcache          VerifiedTxnCache
+	block            bookkeeping.Block
+	proto            config.ConsensusParams
+	verificationPool execpool.BacklogPool
+
+	ctx      context.Context
+	cf       context.CancelFunc
+	txgroups chan []transactions.SignedTxnWithAD
+	done     chan error
+}
+
+func (tv *evalTxValidator) run() {
+	for txgroup := range tv.txgroups {
+		select {
+		case <-tv.ctx.Done():
+			tv.done <- tv.ctx.Err()
+			tv.cf()
+			close(tv.done)
+			return
+		default:
+		}
+		groupNoAD := make([]transactions.SignedTxn, len(txgroup))
+		for i := range txgroup {
+			groupNoAD[i] = txgroup[i].SignedTxn
+		}
+		ctxs := verify.PrepareContexts(groupNoAD, tv.block.BlockHeader)
+
+		for gi, tx := range txgroup {
+			err := validateTransaction(tx.SignedTxn, tv.block, tv.proto, tv.txcache, ctxs[gi], tv.verificationPool)
+			if err != nil {
+				//logging.Base().Errorf("evalTxValidator reject tx %s", tx.ID())
+				tv.done <- err
+				tv.cf()
+				close(tv.done)
+				return
+			}
+		}
+	}
+	close(tv.done)
+}
+
+func validateTransaction(txn transactions.SignedTxn, block bookkeeping.Block, proto config.ConsensusParams, txcache VerifiedTxnCache, ctx verify.Context, verificationPool execpool.BacklogPool) error {
+	// Transaction valid (not expired)?
+	err := txn.Txn.Alive(block)
+	if err != nil {
+		return err
+	}
+
+	// Well-formed on its own?
+	err = txn.Txn.WellFormed(ctx.CurrSpecAddrs, proto)
```

Comment:
`WellFormed` does get called in `verify.TxnPool` below. I am not 100% sure yet if that means it's redundant here

Reply:
now that you mention it, _I'm_ sure. it could go badly if someone stuck a txn in the cache as Verified without running WellFormed on it, so let's never do that. The other two paths into the TransactionPool call verify.TxnPool() or verify.Txn() and those call WellFormed(), so I think we're currently covered.